pageserver/src/page_service.rs

```diff
@@ -447,6 +539,73 @@ impl PageServerHandler {
         Ok(())
     }
 
+    fn handle_import_basebackup(
+        &self,
+        pgb: &mut PostgresBackend,
+        tenant_id: ZTenantId,
+        timeline_id: ZTimelineId,
+        base_lsn: Lsn,
+        _end_lsn: Lsn,
+    ) -> anyhow::Result<()> {
+        thread_mgr::associate_with(Some(tenant_id), Some(timeline_id));
+        let _enter =
+            info_span!("import basebackup", timeline = %timeline_id, tenant = %tenant_id).entered();
+
+        // Create empty timeline
+        info!("creating new timeline");
+        let repo = tenant_mgr::get_repository_for_tenant(tenant_id)?;
+        let timeline = repo.create_empty_timeline(timeline_id, Lsn(0))?;
+        let repartition_distance = repo.get_checkpoint_distance();
+        let mut datadir_timeline =
+            DatadirTimeline::<LayeredRepository>::new(timeline, repartition_distance);
+
+        // TODO mark timeline as not ready until it reaches end_lsn?
+        // This would prevent compute node from connecting to it and writing conflicting wal.
+
+        // TODO leave clean state on error
+
+        // Import basebackup provided via CopyData
+        info!("importing basebackup");
+        pgb.write_message(&BeMessage::CopyInResponse)?;
+        let reader = CopyInReader::new(pgb);
+        import_basebackup_from_tar(&mut datadir_timeline, reader, base_lsn)?;
+
+        info!("done");
```

Comment:
Let's reword this message to something more informative, i.e. `imported timeline {} timeline.lsn {}`

Also, we probably need to call ` timeline.tline.checkpoint(CheckpointConfig::Forced)?;` here to flush the data.

Reply:
We're inside `info_span!`, which provides that information on the same line.

Good point about checkpointing. In the case of import from postgres, we're not done until the data gets to s3.

Off topic: will there be issues with two pageservers writing the same timeline to s3, in the case of "import from other pageserver"?

Reply:
Two concurrent writes into the same timelines are indeed not desirable at the very least.
Alas, there's no sync mechanism to ensure exclusive S3 write in pageserver (yet it's not that hard to implement one via etcd now), but we can do either:

* turn off S3 upload on the new pageserver during the migration period (just omit the S3 settings, basically)
* reproduce the situation manually and see how pageservers would behave, chances are that layers created by different pageservers won't overlap

Reply:
In addition, the new pageserver should never read layer files written by the old pageserver. They have conflicting histories, since the new pageserver thinks the entire database was just created. Not to mention, they likely have conflicting storage formats.

The easiest solution is to write in s3 under a new prefix. For example, prefix the layer object names with storage version number. That prevents migration across pageservers with the same version though (not necessary, but nice to have).

Reply:
I think that we should rather disable/reroute s3 writes for old pageserver instead, because old S3 layers won't be useful after migration.
 
Something like:
1. change old tenant config to `[s3_path].old`
2. move old tenant layers to `[s3_path].old`
3. move data between pageservers. new tenant will write to normal s3 location.

Reply:
Added checkpoint